gnumed/gnumed/client/doc/man-pages/gm-zip+sign_backups.8
.TH GM-ZIP+SIGN_BACKUPS 8 "2011 May 10th" "GNUmed database backup zip+sign"

.SH NAME
.B gm-zip+sign_backups
- pack and sign backups of GNUmed databases

.SH SYNOPSIS
.B gm-zip+sign_backups

.SH DESCRIPTION
The
.B gm-zip+sign_backups
script compresses and signs GNUmed database backups.

It is a really good idea to use it in a daily cron/anacron
job to ensure that GNUmed patient data are safe against data
loss.  The Debian GNU/Linux packaged version of the GNUmed
server installs such a cron job by default.

.SH EXAMPLES

.B anacron

The following line could be added to a system's
/etc/anacrontab to make sure it creates daily
database backups for GNUmed:

1 15 gnumed-<your-company>-sign-backups /usr/bin/gm-zip+sign_backups.sh

.B cron

Add the following line to a crontab file to sign
database backups at 12:47 and 19:47 every day:

47 12,19 * * * * /usr/bin/gm-zip+sign_backups.sh

.B PROCMAIL

It is useful to have a PROCMAIL rule for the digital notary
server replies piping them into the stoarage area where the
backups are kept.


.SH SEE ALSO
gm-backup_database(8)

/etc/gnumed/gnumed-backup.conf

anacron(8)

cron(8)

.SH AUTHOR
This manual page was written by Karsten Hilbert, for the
Debian GNU/Linux system (but may be used by others).


CHANGELOG
*****************************************************************
# client
*****************************************************************

------------------------------------------------
# rel-0-7-patches
------------------------------------------------

	0.7.0

- new: manage diagnostic orgs
- new: manage date of death per patient
- new: manage discontinuation of substance intake
- new: turn substance intake entry into allergy
- new: hooks
	- after_substance_intake_modified
	- hook after_test_result_modified
- new: visual progress notes
- new: initial support for FreeDiams drug database frontend

- improved: login dialag:
	- hint about local DB on failure
	- do not show password ;-)
- improved: show allergies in substance intake edit area
- improved: warn on creating database users with empty passwords
- improved: default region/country for new-person creation
- improved: show lab name/contact details in measurement tooltip
- improved: always use new person creation rather than old
- improved: placeholders:
	- name::*::*
- improved: various typos
- improved: do not fail if unable to show part when adding documents
- improved: cleaner display of documents in tree
- improved: ARRIBA is now v2.0.4
- improved: using Git for SCM now
- improved: gnumed.py now directly under client/

------------------------------------------------
# rel-0-6-patches
------------------------------------------------

	0.6.3

FIX: faulty tooltip on EMR tree items w/ no data to display

	0.6.2

FIX: properly attribute James Busser, MD
FIX: do not fail on attempt to <delete> when no allergy entry selected
FIX: do not fail on tooltipping empty results/meds grids
FIX: fail less haphazardly on invalid dates in wx -> py conversion

	0.6.1

FIX: WRONG most-recent result displayed in measurement type tooltip
FIX: missing .GetValue() on abnormality indicator PRW on updating a test result
FIX: exception on problem with any placeholder rather than continuing with a notice
FIX: work around one more bogus PyDeadObjectError exception
FIX: re-add pt_BR translation to release tarball
FIX: missing dependencies/faulty paths in gm-install_client_locally.sh
FIX: exception on adding health issue from GUI
FIX: several typos

	0.6.0

- can promote an episode to being a health issue
- can add home phone/external ID to newly created patient directly
- can track diagnostic certainty classification (ABCD) on episodes and issues
- can track procedures performed on a patient
- can do end user friendly free-text search across all EMRs
- can move all progress notes of a pre-selected list of encounters to another episode
- can manage provinces
- can manage substance intake
- can print medication list
- can print LaTeX as well as OOo letters
	- referral letter template contributed by Rogerio Luz and James Busser
- can interface with German "MMI/Gelbe Liste" external drug database
	- show info on drug/substance by PZN / name
	- show interactions
	- import drugs/substances
- can display UI in Polish and others (partially)
- can include potential problems in problem list of soap plugin
- can remove DOB from person

- improved (more) placeholders
	- gender to re placement mapper
	- medication list
	- allergies list
	- problem list
- improved inbox

- improved tarballs: include schema/API docs, better names
- improved import path detection

------------------------------------------------
# rel-0-5-patches
------------------------------------------------

	0.5.1 -> 0.5.2

FIX: obsolete LogException in gmMimeLib.py
FIX: failure on adding patient to waiting list with no patient active
FIX: wx.K_* must be wx.WXK_*
NEW: better default size of encounter details editor

	0.5.0 -> 0.5.1

FIX: be more careful when deleting measurement types - they could be in use
FIX: faulty access to "pk" column after creating measurement type from phrasewheel
FIX: missing "import os" in measurement widgets code
FIX: improved starting of new encounter from SOAP widget

NEW: install ARRIBA jar into /usr/local/bin
NEW: man page for gm-install_arriba

	0.5.0

- can track hospital stays
- can group health issues by arbitrary grouping tag
- can display/edit *all* encounters of a patient in a list
- can move individual soap rows between encounters
- can delete progress notes
- can delete workplaces from GUI
- can show measurements encyclopedia from grid
- can manage measurement types
- can list meta measurement types
- can start external MI/stroke risk assessment calculator
- can install ARRIBA MI/stroke risk assessment calculator
- can download and import LOINC data
- can import ATC data
- can disconnect from database welcome message dialog
- can handle persons without DOB
- can display deceased person's age properly
- can remove persons from the database

- new/simpler placeholder style for OOo documents
- protection against %s-count mismatch in _()
- improved display of measurement types in grid
- improved sending of bug reports (log = attachment)
- improved new-patient dialog
- improved sorting of documents in formatted episode summary
- improved "About Database"/startup banner
- improved episode formatting

- has some revision info in EMR Journal view
- has built-in fallback workplace for problem situations

------------------------------------------------
# rel-0-4-patches
------------------------------------------------

	0.4.6 -> 0.4.7

FIX: faulty get_selected_cells() in measurement grid
FIX: failed to properly check problem field in measurements edit area
FIX: missing ) on setting document type translation

	0.4.5 -> 0.4.6

FIX: missing space in update-available message
FIX: search_for_document by type ID: fk_type must be pk_type

	0.4.4 -> 0.4.5

FIX: _log.panic hiccuping on .conf files with malformed lists, should be _log.critical
FIX: on Python 2.6 __new__ doesn't like extra parameters anymore

	0.4.3 -> 0.4.4

FIX: re-enable path sanity check that fell off when fixing Windows
FIX: properly backpatch "properly find last-but-one encounter for an episode or issue"

gnumed/gnumed/server/gm-zip+sign_backups.sh
#!/bin/bash

#==============================================================
# author: Karsten Hilbert
# license: GPL v2
#
# anacron
# -------
#  The following line could be added to a system's
#  /etc/anacrontab to make sure it creates daily
#  database backups for GNUmed:
#
#  1       15      gnumed-<your-company>-sign-backups    /usr/bin/gm-zip+sign_backups.sh
#
#
# cron
# ----
#  Add the following line to a crontab file to sign
#  database backups at 12:47 and 19:47 every day:
#
#  47 12,19 * * * * /usr/bin/gm-zip+sign_backups.sh
#
#
# It is useful to have a PROCMAIL rule for the GNotary server replies
# piping them into the stoarage area where the backups are kept.
#==============================================================

CONF="/etc/gnumed/gnumed-backup.conf"

#==============================================================
# There really should not be any need to
# change anything below this line.
#==============================================================

# load config file
if [ -r ${CONF} ] ; then
	. ${CONF}
else
	echo "Cannot read configuration file ${CONF}. Aborting."
	exit 1
fi

TS=`date +%Y-%m-%d-%H-%M-%S`
BACKUP_BASENAME="backup-${GM_DATABASE}-${INSTANCE_OWNER}"

cd ${BACKUP_DIR}
if test "$?" != "0" ; then
	echo "Cannot change into backup directory [${BACKUP_DIR}]. Aborting."
	exit 1
fi

shopt -s -q nullglob

# zip up any backups
for BACKUP in ${BACKUP_BASENAME}-*.tar ; do

	# are the backup and ...
	TAR_OPEN=`lsof | grep ${BACKUP}`
	# ... the corresponding bz2 both open at the moment ?
	BZ2_OPEN=`lsof | grep ${BACKUP}.bz2`
	if test -z "${TAR_OPEN}" -a -z "${BZ2_OPEN}" ; then
		# no: remove the bz2 and start over compressing
		rm -f ${BACKUP}.bz2
	else
		# yes: skip to next backup
		continue
	fi

	bzip2 -zq -${COMPRESSION_LEVEL} ${BACKUP}
	bzip2 -tq ${BACKUP}.bz2
	# FIXME: add check for exit code

	chmod ${BACKUP_MASK} ${BACKUP}.bz2
	chown ${BACKUP_OWNER} ${BACKUP}.bz2

	# Reed-Solomon error protection support
#	if test -n ${ADD_ECC} ; then
#		rsbep
#	fi

	# GNotary support
	if test -n ${GNOTARY_TAN} ; then
		LOCAL_MAILER=`which mail`

		#SHA512="SHA 512:"`sha512sum -b ${BACKUP_FILENAME}.tar.bz2`
		SHA512=`openssl dgst -sha512 -hex ${BACKUP}.bz2`
		RMD160=`openssl dgst -ripemd160 -hex ${BACKUP}.bz2`

		export REPLYTO=${SIG_RECEIVER}

		# send mail
		(
			echo " "
			echo "<?xml version=\"1.0\" encoding=\"iso-8859-1\" ?>"
			echo "<message>"
			echo "	<tan>$GNOTARY_TAN</tan>"
			echo "	<action>notarize</action>"
			echo "	<hashes number=\"2\">"
			echo "		<hash file=\"${BACKUP}.bz2\" modified=\"${TS}\" algorithm=\"SHA-512\">${SHA512}</hash>"
			echo "		<hash file=\"${BACKUP}.bz2\" modified=\"${TS}\" algorithm=\"RIPE-MD-160\">${RMD160}</hash>"
			echo "	</hashes>"
			echo "</message>"
			echo " "
		) | $LOCAL_MAILER -s "gnotarize" $GNOTARY_SERVER
	fi

done

exit 0

#==============================================================
